clippy_lints/src/redundant_closure_call.rs

```diff
@@ -51,59 +51,136 @@ impl ReturnVisitor {
     }
 }
 
-impl<'ast> ast_visit::Visitor<'ast> for ReturnVisitor {
-    fn visit_expr(&mut self, ex: &'ast ast::Expr) {
-        if let ast::ExprKind::Ret(_) | ast::ExprKind::Try(_) = ex.kind {
+impl<'tcx> Visitor<'tcx> for ReturnVisitor {
+    fn visit_expr(&mut self, ex: &'tcx hir::Expr<'tcx>) {
+        if let hir::ExprKind::Ret(_) | hir::ExprKind::Match(.., hir::MatchSource::TryDesugar) = ex.kind {
             self.found_return = true;
         }
 
-        ast_visit::walk_expr(self, ex);
+        hir_visit::walk_expr(self, ex);
     }
 }
 
-impl EarlyLintPass for RedundantClosureCall {
-    fn check_expr(&mut self, cx: &EarlyContext<'_>, expr: &ast::Expr) {
+/// Checks if the body is owned by an async closure
+fn is_async_closure(body: &hir::Body<'_>) -> bool {
+    if let hir::ExprKind::Closure(closure) = body.value.kind
+        && let [resume_ty] = closure.fn_decl.inputs
+        && let hir::TyKind::Path(hir::QPath::LangItem(hir::LangItem::ResumeTy, ..)) = resume_ty.kind
+    {
+        true
+    } else {
+        false
+    }
+}
+
+/// Tries to find the innermost closure:
+/// ```rust,ignore
+/// (|| || || || 42)()()()()
+///  ^^^^^^^^^^^^^^          given this nested closure expression
+///           ^^^^^          we want to return this closure
+/// ```
+/// It also has a parameter for how many steps to go in at most, so as to
+/// not take more closures than there are calls.
+fn find_innermost_closure<'tcx>(
+    cx: &LateContext<'tcx>,
+    mut expr: &'tcx hir::Expr<'tcx>,
+    mut steps: usize,
+) -> Option<(&'tcx hir::Expr<'tcx>, &'tcx hir::FnDecl<'tcx>, hir::IsAsync)> {
+    let mut data = None;
+
+    while let hir::ExprKind::Closure(closure) = expr.kind
+        && let body = cx.tcx.hir().body(closure.body)
+        && {
+            let mut visitor = ReturnVisitor::new();
+            visitor.visit_expr(body.value);
+            !visitor.found_return
+        }
+        && steps > 0
+    {
+        expr = body.value;
+        data = Some((body.value, closure.fn_decl, if is_async_closure(body) {
+            hir::IsAsync::Async
+        } else {
+            hir::IsAsync::NotAsync
+        }));
+        steps -= 1;
+    }
+
+    data
+}
+
+/// "Walks up" the chain of calls to find the outermost call expression, and returns the depth:
+/// ```rust,ignore
+/// (|| || || 3)()()()
+///             ^^      this is the call expression we were given
+///                 ^^  this is what we want to return (and the depth is 3)
+/// ```
+fn get_parent_call_exprs<'tcx>(
+    cx: &LateContext<'tcx>,
+    mut expr: &'tcx hir::Expr<'tcx>,
+) -> (&'tcx hir::Expr<'tcx>, usize) {
+    let mut depth = 1;
+    while let Some(parent) = get_parent_expr(cx, expr)
+        && let hir::ExprKind::Call(recv, _) = parent.kind
+        && let hir::ExprKind::Call(..) = recv.kind
```

Comment:
I think this is needed, because without it, this code:
```rs
fn foo(_: i32, _: i32) {}

foo((|| || 42)()(), 5);
```
gets this suggestion:
```
warning: try not to call a closure in the expression where it is declared
   --> test2.rs:150:5
    |
150 |     foo((|| || 42)()(), 5);
    |     ^^^^^^^^^^^^^^^^^^^^^^ help: try doing something like: `42`
```
that is, it includes the `foo()` call, when it should have stopped at the last closure call.
This happens because in that snippet, the parent of `(|| || 42)()` is the `foo(...)` call. That extra check works avoids this because it makes sure that the receiver of `foo()` is a function call, which, here it isn't (it's a path to `foo`).

This deserves a test case though...

Reply:
Yeah I changed it in my local checkout of the PR and it all seemed to work so I suggested it. Thanks for telling!

Reply:
Okay, sorry for the confusion and noise, I think I have a better fix for this. 
<details>
<summary> Explanation <sub>(this might confuse you even more, so open at your own risk)</sub> </summary>

Instead of checking if the receiver is a call expression (which clearly fails in some cases as shown), we just need to check if the receiver matches with the expression currently being checked, which we can just do by comparing the spans. Example:

Say we are checking this expression:
```rs
bar()((|| || 42)()(), 5);
      ^^^^^^^^^^^^ we are currently processing this expression
```
First, check that the parent (here: `(|| || 42)()()` is a call expression (it is) and also check that the receiver (`(|| || 42)()`)is the same as the expression we are checking, and here they are the same.

Now going up the parent chain:
```rs

bar()((|| || 42)()(), 5);
      ^^^^^^^^^^^^^^ 
```
Now we check that the parent (here: `bar()(....)`) is a call expression (here it is) and also check that the receiver has the same span as the expression. And *that* fails because `bar()` does not have the same span as `(|| || 42)()()`.
</details>